// @nrwl/workspace/src/tasks-runner/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unparse = exports.getOutputsForTargetAndConfiguration = exports.getOutputs = exports.getDependencyConfigs = exports.getDefaultDependencyConfigs = exports.getCommand = exports.getCommandAsString = void 0;
const flat_1 = require("flat");
const output_1 = require("../utilities/output");
const commonCommands = ['build', 'test', 'lint', 'e2e', 'deploy'];
function getCommandAsString(cliCommand, isYarn, task) {
    return getCommand(cliCommand, isYarn, task).join(' ').trim();
}
exports.getCommandAsString = getCommandAsString;
function getCommand(cliCommand, isYarn, task) {
    const args = Object.entries(task.overrides || {}).map(([prop, value]) => `--${prop}=${value}`);
    if (commonCommands.includes(task.target.target)) {
        const config = task.target.configuration
            ? [`--configuration`, task.target.configuration]
            : [];
        return [
            cliCommand,
            ...(isYarn ? [] : ['--']),
            task.target.target,
            task.target.project,
            ...config,
            ...args,
        ];
    }
    else {
        const config = task.target.configuration
            ? `:${task.target.configuration} `
            : '';
        return [
            cliCommand,
            ...(isYarn ? [] : ['--']),
            'run',
            `${task.target.project}:${task.target.target}${config}`,
            ...args,
        ];
    }
}
exports.getCommand = getCommand;
function getDefaultDependencyConfigs(nxJson, runnerOptions) {
    var _a, _b;
    const defaults = (_a = nxJson.targetDependencies) !== null && _a !== void 0 ? _a : {};
    const strictlyOrderedTargets = runnerOptions
        ? (_b = runnerOptions.strictlyOrderedTargets) !== null && _b !== void 0 ? _b : ['build']
        : [];
    // Strictly Ordered Targets depend on their dependencies
    for (const target of strictlyOrderedTargets) {
        defaults[target] = defaults[target] || [];
        defaults[target].push({
            target,
            projects: 'dependencies',
        });
    }
    return defaults;
}
exports.getDefaultDependencyConfigs = getDefaultDependencyConfigs;
function getDependencyConfigs({ project, target }, defaultDependencyConfigs, projectGraph) {
    var _a, _b, _c, _d;
    // DependencyConfigs configured in workspace.json override configurations at the root.
    const dependencyConfigs = (_d = (_c = (_b = (_a = projectGraph.nodes[project].data) === null || _a === void 0 ? void 0 : _a.targets[target]) === null || _b === void 0 ? void 0 : _b.dependsOn) !== null && _c !== void 0 ? _c : defaultDependencyConfigs[target]) !== null && _d !== void 0 ? _d : [];
    for (const dependencyConfig of dependencyConfigs) {
        if (dependencyConfig.projects !== 'dependencies' &&
            dependencyConfig.projects !== 'self') {
            output_1.output.error({
                title: `dependsOn is improperly configured for ${project}:${target}`,
                bodyLines: [
                    `dependsOn.projects is ${dependencyConfig.projects} but should be "self" or "dependencies"`,
                ],
            });
            process.exit(1);
        }
    }
    return dependencyConfigs;
}
exports.getDependencyConfigs = getDependencyConfigs;
function getOutputs(p, task) {
    return getOutputsForTargetAndConfiguration(task, p[task.target.project]);
}
exports.getOutputs = getOutputs;
function getOutputsForTargetAndConfiguration(task, node) {
    var _a;
    const { target, configuration } = task.target;
    const targets = node.data.targets[target];
    const options = Object.assign(Object.assign(Object.assign({}, targets.options), (_a = targets === null || targets === void 0 ? void 0 : targets.configurations) === null || _a === void 0 ? void 0 : _a[configuration]), task.overrides);
    if (targets === null || targets === void 0 ? void 0 : targets.outputs) {
        return targets.outputs.map((output) => interpolateOutputs(output, options));
    }
    // Keep backwards compatibility in case `outputs` doesn't exist
    if (options.outputPath) {
        return Array.isArray(options.outputPath)
            ? options.outputPath
            : [options.outputPath];
    }
    else if (target === 'build' || target === 'prepare') {
        return [
            `dist/${node.data.root}`,
            `${node.data.root}/dist`,
            `${node.data.root}/build`,
            `${node.data.root}/public`,
        ];
    }
    else {
        return [];
    }
}
exports.getOutputsForTargetAndConfiguration = getOutputsForTargetAndConfiguration;
function unparse(options) {
    const unparsed = [];
    for (const key of Object.keys(options)) {
        const value = options[key];
        unparseOption(key, value, unparsed);
    }
    return unparsed;
}
exports.unparse = unparse;
function unparseOption(key, value, unparsed) {
    if (value === true) {
        unparsed.push(`--${key}`);
    }
    else if (value === false) {
        unparsed.push(`--no-${key}`);
    }
    else if (Array.isArray(value)) {
        value.forEach((item) => unparseOption(key, item, unparsed));
    }
    else if (Object.prototype.toString.call(value) === '[object Object]') {
        const flattened = flat_1.flatten(value, { safe: true });
        for (const flattenedKey in flattened) {
            unparseOption(`${key}.${flattenedKey}`, flattened[flattenedKey], unparsed);
        }
    }
    else if (typeof value === 'string' && value.includes(' ')) {
        unparsed.push(`--${key}="${value}"`);
    }
    else if (value != null) {
        unparsed.push(`--${key}=${value}`);
    }
}
function interpolateOutputs(template, data) {
    return template.replace(/{([\s\S]+?)}/g, (match) => {
        let value = data;
        let path = match.slice(1, -1).trim().split('.').slice(1);
        for (let idx = 0; idx < path.length; idx++) {
            if (!value[path[idx]]) {
                throw new Error(`Could not interpolate output {${match}}!`);
            }
            value = value[path[idx]];
        }
        return value;
    });
}
