// @nrwl/angular/src/builders/ng-packagr-lite/ng-packagr-adjustments/entry-point.di.js
"use strict";
/**
 * Wires everything together and provides it to the DI system, taking what
 * we still want from the original ng-packagr library and replacing those
 * where Nx takes over with Nx specific functions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NX_ENTRY_POINT_PROVIDERS = exports.NX_ENTRY_POINT_TRANSFORM = exports.NX_ENTRY_POINT_TRANSFORM_TOKEN = void 0;
const injection_js_1 = require("injection-js");
const transform_di_1 = require("ng-packagr/lib/graph/transform.di");
const compile_ngc_1 = require("./compile-ngc");
const entry_point_1 = require("./entry-point");
const write_package_di_1 = require("./write-package.di");
exports.NX_ENTRY_POINT_TRANSFORM_TOKEN = new injection_js_1.InjectionToken(`nx.v1.entryPointTransform`);
exports.NX_ENTRY_POINT_TRANSFORM = transform_di_1.provideTransform({
    provide: exports.NX_ENTRY_POINT_TRANSFORM_TOKEN,
    useFactory: entry_point_1.nxEntryPointTransformFactory,
    deps: [compile_ngc_1.NX_COMPILE_NGC_TOKEN, write_package_di_1.NX_WRITE_PACKAGE_TRANSFORM_TOKEN],
});
exports.NX_ENTRY_POINT_PROVIDERS = [
    exports.NX_ENTRY_POINT_TRANSFORM,
    compile_ngc_1.NX_COMPILE_NGC_TRANSFORM,
    write_package_di_1.NX_WRITE_PACKAGE_TRANSFORM,
];
