// @nrwl/angular/src/builders/ng-packagr-lite/ng-packagr-adjustments/write-package.js
"use strict";
/**
 * Adapted from original ng-packagr
 *
 * Change the package.json metadata to only use
 * the ESM2015 output as it's the only one generated.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.nxWritePackageTransform = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const node_1 = require("ng-packagr/lib/graph/node");
const transform_1 = require("ng-packagr/lib/graph/transform");
const nodes_1 = require("ng-packagr/lib/ng-package/nodes");
const fs_1 = require("ng-packagr/lib/utils/fs");
const glob_1 = require("ng-packagr/lib/utils/glob");
const path_1 = require("ng-packagr/lib/utils/path");
const path = require("path");
exports.nxWritePackageTransform = transform_1.transformFromPromise((graph) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    const entryPoint = graph.find(nodes_1.isEntryPointInProgress());
    const ngEntryPoint = entryPoint.data.entryPoint;
    const ngPackageNode = graph.find(nodes_1.isPackage);
    const ngPackage = ngPackageNode.data;
    const { destinationFiles } = entryPoint.data;
    const ignorePaths = [
        '.gitkeep',
        '**/.DS_Store',
        '**/Thumbs.db',
        '**/node_modules/**',
        `${ngPackage.dest}/**`,
    ];
    if (ngPackage.assets.length && !ngEntryPoint.isSecondaryEntryPoint) {
        const assetFiles = [];
        // COPY ASSET FILES TO DESTINATION
        devkit_1.logger.log('Copying assets');
        try {
            for (let asset of ngPackage.assets) {
                asset = path.join(ngPackage.src, asset);
                if (yield fs_1.exists(asset)) {
                    const stats = yield fs_1.stat(asset);
                    if (stats.isFile()) {
                        assetFiles.push(asset);
                        continue;
                    }
                    if (stats.isDirectory()) {
                        asset = path.join(asset, '**/*');
                    }
                }
                const files = yield glob_1.globFiles(asset, {
                    ignore: ignorePaths,
                    cache: ngPackageNode.cache.globCache,
                    dot: true,
                    nodir: true,
                });
                if (files.length) {
                    assetFiles.push(...files);
                }
            }
            for (const file of assetFiles) {
                const relativePath = path.relative(ngPackage.src, file);
                const destination = path.resolve(ngPackage.dest, relativePath);
                const nodeUri = nodes_1.fileUrl(path_1.ensureUnixPath(file));
                let node = graph.get(nodeUri);
                if (!node) {
                    node = new node_1.Node(nodeUri);
                    graph.put(node);
                }
                entryPoint.dependsOn(node);
                yield fs_1.copyFile(file, destination);
            }
        }
        catch (error) {
            throw error;
        }
    }
    // 6. WRITE PACKAGE.JSON
    try {
        devkit_1.logger.info('Writing package metadata');
        const relativeUnixFromDestPath = (filePath) => path_1.ensureUnixPath(path.relative(ngEntryPoint.destinationPath, filePath));
        const { enableIvy, compilationMode } = entryPoint.data.tsConfig.options;
        yield writePackageJson(ngEntryPoint, ngPackage, {
            module: relativeUnixFromDestPath(destinationFiles.esm2015),
            esm2015: relativeUnixFromDestPath(destinationFiles.esm2015),
            typings: relativeUnixFromDestPath(destinationFiles.declarations),
            // Ivy doesn't generate metadata files
            metadata: enableIvy
                ? undefined
                : relativeUnixFromDestPath(destinationFiles.metadata),
            // webpack v4+ specific flag to enable advanced optimizations and code splitting
            sideEffects: ngEntryPoint.sideEffects,
        }, !!enableIvy, compilationMode);
    }
    catch (error) {
        throw error;
    }
    devkit_1.logger.info(`Built ${ngEntryPoint.moduleId}`);
    return graph;
}));
/**
 * Creates and writes a `package.json` file of the entry point used by the `node_module`
 * resolution strategies.
 *
 * #### Example
 *
 * A consumer of the entry point depends on it by `import {..} from '@my/module/id';`.
 * The module id `@my/module/id` will be resolved to the `package.json` file that is written by
 * this build step.
 * The properties `main`, `module`, `typings` (and so on) in the `package.json` point to the
 * flattened JavaScript bundles, type definitions, (...).
 *
 * @param entryPoint An entry point of an Angular package / library
 * @param additionalProperties Additional properties, e.g. binary artefacts (bundle files), to merge into `package.json`
 */
function writePackageJson(entryPoint, pkg, additionalProperties, isIvy, compilationMode) {
    var _a, _b, _c;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        // set additional properties
        const packageJson = Object.assign(Object.assign({}, entryPoint.packageJson), additionalProperties);
        // read tslib version from `@angular/compiler` so that our tslib
        // version at least matches that of angular if we use require('tslib').version
        // it will get what installed and not the minimum version nor if it is a `~` or `^`
        // this is only required for primary
        if (!entryPoint.isSecondaryEntryPoint) {
            if (!((_a = packageJson.peerDependencies) === null || _a === void 0 ? void 0 : _a.tslib) &&
                !((_b = packageJson.dependencies) === null || _b === void 0 ? void 0 : _b.tslib)) {
                const { peerDependencies: angularPeerDependencies = {}, dependencies: angularDependencies = {}, } = require('@angular/compiler/package.json');
                const tsLibVersion = angularPeerDependencies.tslib || angularDependencies.tslib;
                if (tsLibVersion) {
                    packageJson.dependencies = Object.assign(Object.assign({}, packageJson.dependencies), { tslib: tsLibVersion });
                }
            }
            else if ((_c = packageJson.peerDependencies) === null || _c === void 0 ? void 0 : _c.tslib) {
                devkit_1.logger.warn(`'tslib' is no longer recommended to be used as a 'peerDependencies'. Moving it to 'dependencies'.`);
                packageJson.dependencies = Object.assign(Object.assign({}, (packageJson.dependencies || {})), { tslib: packageJson.peerDependencies.tslib });
                delete packageJson.peerDependencies.tslib;
            }
        }
        // Verify non-peerDependencies as they can easily lead to duplicate installs or version conflicts
        // in the node_modules folder of an application
        const allowedList = pkg.allowedNonPeerDependencies.map((value) => new RegExp(value));
        try {
            checkNonPeerDependencies(packageJson, 'dependencies', allowedList);
        }
        catch (e) {
            yield fs_1.rmdir(entryPoint.destinationPath, { recursive: true });
            throw e;
        }
        // Removes scripts from package.json after build
        if (packageJson.scripts) {
            if (pkg.keepLifecycleScripts !== true) {
                devkit_1.logger.info(`Removing scripts section in package.json as it's considered a potential security vulnerability.`);
                delete packageJson.scripts;
            }
            else {
                devkit_1.logger.warn(`You enabled keepLifecycleScripts explicitly. The scripts section in package.json will be published to npm.`);
            }
        }
        if (isIvy &&
            !entryPoint.isSecondaryEntryPoint &&
            compilationMode !== 'partial') {
            const scripts = packageJson.scripts || (packageJson.scripts = {});
            scripts.prepublishOnly =
                'node --eval "console.error(\'' +
                    'ERROR: Trying to publish a package that has been compiled by Ivy in full compilation mode. This is not allowed.\\n' +
                    'Please delete and rebuild the package with Ivy partial compilation mode, before attempting to publish.\\n' +
                    '\')" ' +
                    '&& exit 1';
        }
        // keep the dist package.json clean
        // this will not throw if ngPackage field does not exist
        delete packageJson.ngPackage;
        const packageJsonPropertiesToDelete = [
            'stylelint',
            'prettier',
            'browserslist',
            'devDependencies',
            'jest',
            'workspaces',
            'husky',
        ];
        for (const prop of packageJsonPropertiesToDelete) {
            if (prop in packageJson) {
                delete packageJson[prop];
                devkit_1.logger.info(`Removing ${prop} section in package.json.`);
            }
        }
        packageJson.name = entryPoint.moduleId;
        yield fs_1.writeFile(path.join(entryPoint.destinationPath, 'package.json'), JSON.stringify(packageJson, undefined, 2));
    });
}
function checkNonPeerDependencies(packageJson, property, allowed) {
    if (!packageJson[property]) {
        return;
    }
    for (const dep of Object.keys(packageJson[property])) {
        if (!allowed.some((regex) => regex.test(dep))) {
            devkit_1.logger.warn(`Distributing npm packages with '${property}' is not recommended. Please consider adding ${dep} to 'peerDependencies' or remove it from '${property}'.`);
            throw new Error(`Dependency ${dep} must be explicitly allowed using the "allowedNonPeerDependencies" option.`);
        }
    }
}
