// @nrwl/angular/bundles/nrwl-angular-testing.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('jasmine-marbles'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@nrwl/angular/testing', ['exports', 'jasmine-marbles', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.nrwl = global.nrwl || {}, global.nrwl.angular = global.nrwl.angular || {}, global.nrwl.angular.testing = {}), global['jasmine-marbles'], global.rxjs.operators));
}(this, (function (exports, jasmineMarbles, operators) { 'use strict';

    /**
     * @whatItDoes reads all the values from an observable and returns a promise
     * with an array of all values. This should be used in combination with async/await.
     *
     * ## Example
     *
     * ```typescript
     * const obs = of(1, 2, 3, 4);
     * const res = await readAll(obs)
     * expect(res).toEqual([1, 2, 3, 4]);
     * ```
     */
    function readAll(o) {
        return o.pipe(operators.toArray()).toPromise();
    }
    /**
     * @whatItDoes reads the first value from an observable and returns a promise
     * with it. This should be used in combination with async/await.
     *
     * ## Example
     *
     * ```typescript
     * const obs = of(1, 2, 3, 4);
     * const res = await readFirst(obs)
     * expect(res).toEqual(1);
     * ```
     */
    function readFirst(o) {
        return o.pipe(operators.first()).toPromise();
    }

    /**
     * Generated bundle index. Do not edit.
     */

    Object.defineProperty(exports, 'cold', {
        enumerable: true,
        get: function () {
            return jasmineMarbles.cold;
        }
    });
    Object.defineProperty(exports, 'getTestScheduler', {
        enumerable: true,
        get: function () {
            return jasmineMarbles.getTestScheduler;
        }
    });
    Object.defineProperty(exports, 'hot', {
        enumerable: true,
        get: function () {
            return jasmineMarbles.hot;
        }
    });
    Object.defineProperty(exports, 'time', {
        enumerable: true,
        get: function () {
            return jasmineMarbles.time;
        }
    });
    exports.readAll = readAll;
    exports.readFirst = readFirst;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
