// @nrwl/angular/src/schematics/upgrade-module/upgrade-module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const path = require("path");
const workspace_1 = require("@nrwl/workspace");
const upgrade_1 = require("../../utils/upgrade");
const workspace_2 = require("@nrwl/workspace");
const ast_utils_1 = require("../../utils/ast-utils");
const ast_utils_2 = require("@nrwl/workspace/src/utils/ast-utils");
const devkit_1 = require("@nrwl/devkit");
function addImportsToModule(options) {
    return (host) => {
        const { moduleClassName, modulePath, moduleSource } = ast_utils_1.readBootstrapInfo(host, options.project);
        workspace_1.insert(host, modulePath, [
            ast_utils_2.insertImport(moduleSource, modulePath, `configure${devkit_1.names(options.name).className}, upgradedComponents`, `../${devkit_1.names(options.name).fileName}-setup`),
            ast_utils_2.insertImport(moduleSource, modulePath, 'UpgradeModule', '@angular/upgrade/static'),
            ...ast_utils_1.addImportToModule(moduleSource, modulePath, 'UpgradeModule'),
            ...ast_utils_1.addDeclarationToModule(moduleSource, modulePath, '...upgradedComponents'),
            ...ast_utils_1.addEntryComponents(moduleSource, modulePath, ast_utils_1.getBootstrapComponent(moduleSource, moduleClassName)),
        ]);
        return host;
    };
}
function addNgDoBootstrapToModule(options) {
    return (host) => {
        const { moduleClassName, modulePath, moduleSource } = ast_utils_1.readBootstrapInfo(host, options.project);
        workspace_1.insert(host, modulePath, [
            ...workspace_1.addParameterToConstructor(moduleSource, modulePath, {
                className: moduleClassName,
                param: 'private upgrade: UpgradeModule',
            }),
            ...workspace_1.addMethod(moduleSource, modulePath, {
                className: moduleClassName,
                methodHeader: 'ngDoBootstrap(): void',
                body: `
configure${devkit_1.names(options.name).className}(this.upgrade.injector);
this.upgrade.bootstrap(document.body, ['downgraded', '${options.name}']);
        `,
            }),
            ...ast_utils_1.removeFromNgModule(moduleSource, modulePath, 'bootstrap'),
        ]);
        return host;
    };
}
function createFiles(angularJsImport, options) {
    return (host, context) => {
        const { moduleClassName, mainPath, moduleSpec, bootstrapComponentClassName, bootstrapComponentFileName, } = ast_utils_1.readBootstrapInfo(host, options.project);
        const dir = path.dirname(mainPath);
        const templateSource = schematics_1.apply(schematics_1.url('./files'), [
            schematics_1.template(Object.assign(Object.assign(Object.assign({}, options), { tmpl: '', moduleFileName: moduleSpec, moduleClassName,
                angularJsImport, angularJsModule: options.name, bootstrapComponentClassName,
                bootstrapComponentFileName }), devkit_1.names(options.name))),
            schematics_1.move(dir),
        ]);
        const r = schematics_1.branchAndMerge(schematics_1.chain([schematics_1.mergeWith(templateSource)]));
        return r(host, context);
    };
}
function default_1(options) {
    var _a;
    const angularJsImport = (_a = options.angularJsImport) !== null && _a !== void 0 ? _a : options.name;
    return schematics_1.chain([
        createFiles(angularJsImport, options),
        addImportsToModule(options),
        addNgDoBootstrapToModule(options),
        options.skipPackageJson ? schematics_1.noop() : upgrade_1.addUpgradeToPackageJson(),
        workspace_2.formatFiles(options),
    ]);
}
exports.default = default_1;
