// @nrwl/workspace/src/core/hasher/hasher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hasher = void 0;
const tslib_1 = require("tslib");
const file_utils_1 = require("../file-utils");
const child_process_1 = require("child_process");
const file_hasher_1 = require("./file-hasher");
const hashing_impl_1 = require("./hashing-impl");
const minimatch = require("minimatch");
const perf_hooks_1 = require("perf_hooks");
const stripJsonComments = require("strip-json-comments");
const fileutils_1 = require("../../utilities/fileutils");
class Hasher {
    constructor(projectGraph, nxJson, options, hashing = undefined) {
        this.projectGraph = projectGraph;
        this.nxJson = nxJson;
        this.options = options;
        if (!hashing) {
            this.hashing = hashing_impl_1.defaultHashing;
            this.fileHasher = file_hasher_1.defaultFileHasher;
        }
        else {
            // this is only used for testing
            this.hashing = hashing;
            this.fileHasher = new file_hasher_1.FileHasher(hashing);
            this.fileHasher.clear();
        }
        this.projectHashes = new ProjectHasher(this.projectGraph, this.hashing);
    }
    hashTaskWithDepsAndContext(task) {
        var _a, _b, _c;
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const command = this.hashing.hashArray([
                (_a = task.target.project) !== null && _a !== void 0 ? _a : '',
                (_b = task.target.target) !== null && _b !== void 0 ? _b : '',
                (_c = task.target.configuration) !== null && _c !== void 0 ? _c : '',
                JSON.stringify(task.overrides),
            ]);
            const values = (yield Promise.all([
                this.projectHashes.hashProject(task.target.project, [
                    task.target.project,
                ]),
                this.implicitDepsHash(),
                this.runtimeInputsHash(),
            ]));
            const value = this.hashing.hashArray([
                Hasher.version,
                command,
                ...values.map((v) => v.value),
            ]);
            return {
                value,
                details: {
                    command,
                    sources: values[0].sources,
                    implicitDeps: values[1].sources,
                    runtime: values[2].runtime,
                },
            };
        });
    }
    hashContext() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const values = (yield Promise.all([
                this.implicitDepsHash(),
                this.runtimeInputsHash(),
            ]));
            return {
                implicitDeps: values[0],
                runtime: values[1],
            };
        });
    }
    hashSource(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return this.projectHashes.hashProjectNodeSource(task.target.project);
        });
    }
    hashArray(values) {
        return this.hashing.hashArray(values);
    }
    runtimeInputsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.runtimeInputs)
                return this.runtimeInputs;
            perf_hooks_1.performance.mark('hasher:runtime inputs hash:start');
            this.runtimeInputs = new Promise((res, rej) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const inputs = this.options && this.options.runtimeCacheInputs
                    ? this.options.runtimeCacheInputs
                    : [];
                if (inputs.length > 0) {
                    try {
                        const values = (yield Promise.all(inputs.map((input) => new Promise((res, rej) => {
                            child_process_1.exec(input, (err, stdout, stderr) => {
                                if (err) {
                                    rej(err);
                                }
                                else {
                                    res({ input, value: `${stdout}${stderr}`.trim() });
                                }
                            });
                        }))));
                        const value = this.hashing.hashArray(values.map((v) => v.value));
                        const runtime = values.reduce((m, c) => ((m[c.input] = c.value), m), {});
                        perf_hooks_1.performance.mark('hasher:runtime inputs hash:end');
                        perf_hooks_1.performance.measure('hasher:runtime inputs hash', 'hasher:runtime inputs hash:start', 'hasher:runtime inputs hash:end');
                        res({ value, runtime });
                    }
                    catch (e) {
                        rej(new Error(`Nx failed to execute runtimeCacheInputs defined in nx.json failed:\n${e.message}`));
                    }
                }
                else {
                    res({ value: '', runtime: {} });
                }
            }));
            return this.runtimeInputs;
        });
    }
    implicitDepsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.implicitDependencies)
                return this.implicitDependencies;
            perf_hooks_1.performance.mark('hasher:implicit deps hash:start');
            this.implicitDependencies = new Promise((res) => {
                var _a, _b;
                const implicitDeps = Object.keys((_a = this.nxJson.implicitDependencies) !== null && _a !== void 0 ? _a : {});
                const filesWithoutPatterns = implicitDeps.filter((p) => p.indexOf('*') === -1);
                const patterns = implicitDeps.filter((p) => p.indexOf('*') !== -1);
                const implicitDepsFromPatterns = patterns.length > 0
                    ? ((_b = this.projectGraph.allWorkspaceFiles) !== null && _b !== void 0 ? _b : [])
                        .filter((f) => !!patterns.find((pattern) => minimatch(f.file, pattern)))
                        .map((f) => f.file)
                    : [];
                const fileNames = [
                    ...filesWithoutPatterns,
                    ...implicitDepsFromPatterns,
                    //TODO: vsavkin move the special cases into explicit ts support
                    'tsconfig.base.json',
                    'package-lock.json',
                    'yarn.lock',
                    'pnpm-lock.yaml',
                ];
                const fileHashes = [
                    ...fileNames.map((file) => {
                        const hash = this.fileHasher.hashFile(file);
                        return { file, hash };
                    }),
                    ...this.hashGlobalConfig(),
                ];
                const combinedHash = this.hashing.hashArray(fileHashes.map((v) => v.hash));
                perf_hooks_1.performance.mark('hasher:implicit deps hash:end');
                perf_hooks_1.performance.measure('hasher:implicit deps hash', 'hasher:implicit deps hash:start', 'hasher:implicit deps hash:end');
                res({
                    value: combinedHash,
                    sources: fileHashes.reduce((m, c) => ((m[c.file] = c.hash), m), {}),
                });
            });
            return this.implicitDependencies;
        });
    }
    hashGlobalConfig() {
        return [
            {
                hash: this.fileHasher.hashFile('nx.json', (file) => {
                    try {
                        const r = JSON.parse(stripJsonComments(file));
                        delete r.projects;
                        return JSON.stringify(r);
                    }
                    catch (_a) {
                        return '';
                    }
                }),
                file: 'nx.json',
            },
        ];
    }
}
exports.Hasher = Hasher;
Hasher.version = '2.0';
class ProjectHasher {
    constructor(projectGraph, hashing) {
        this.projectGraph = projectGraph;
        this.hashing = hashing;
        this.sourceHashes = {};
        this.workspaceJson = this.readConfigFile(file_utils_1.workspaceFileName());
        this.nxJson = this.readConfigFile('nx.json');
    }
    hashProject(projectName, visited) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return Promise.resolve().then(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
                var _a;
                const deps = (_a = this.projectGraph.dependencies[projectName]) !== null && _a !== void 0 ? _a : [];
                const depHashes = (yield Promise.all(deps.map((d) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    if (visited.indexOf(d.target) > -1) {
                        return null;
                    }
                    else {
                        visited.push(d.target);
                        return yield this.hashProject(d.target, visited);
                    }
                })))).filter((r) => !!r);
                const projectHash = yield this.hashProjectNodeSource(projectName);
                const sources = depHashes.reduce((m, c) => {
                    return Object.assign(Object.assign({}, m), c.sources);
                }, { [projectName]: projectHash });
                const value = this.hashing.hashArray([
                    ...depHashes.map((d) => d.value),
                    projectHash,
                ]);
                return { value, sources };
            }));
        });
    }
    hashProjectNodeSource(projectName) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.sourceHashes[projectName]) {
                this.sourceHashes[projectName] = new Promise((res) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    var _a, _b;
                    const p = this.projectGraph.nodes[projectName];
                    const fileNames = p.data.files.map((f) => f.file);
                    const values = p.data.files.map((f) => f.hash);
                    const workspaceJson = JSON.stringify((_a = this.workspaceJson.projects[projectName]) !== null && _a !== void 0 ? _a : '');
                    const nxJson = JSON.stringify((_b = this.nxJson.projects[projectName]) !== null && _b !== void 0 ? _b : '');
                    res(this.hashing.hashArray([
                        ...fileNames,
                        ...values,
                        workspaceJson,
                        nxJson,
                    ]));
                }));
            }
            return this.sourceHashes[projectName];
        });
    }
    readConfigFile(path) {
        var _a;
        try {
            const res = fileutils_1.readJsonFile(path);
            (_a = res.projects) !== null && _a !== void 0 ? _a : (res.projects = {});
            return res;
        }
        catch (_b) {
            return { projects: {} };
        }
    }
}
