// @nrwl/angular/src/schematics/ngrx/ngrx.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ngrxGenerator = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const path = require("path");
const rules_1 = require("./rules");
const workspace_1 = require("@nrwl/workspace");
const devkit_1 = require("@nrwl/devkit");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
/**
 * Rule to generate the Nx 'ngrx' Collection
 * Note: see https://nx.dev/latest/angular/guides/misc-ngrx for guide to generated files
 */
function generateNgrxCollection(_options) {
    return (host, context) => {
        const options = normalizeOptions(_options);
        if (!options.module) {
            throw new Error(`The required --module option must be passed`);
        }
        else if (!host.exists(options.module)) {
            throw new Error(`Path does not exist: ${options.module}`);
        }
        const requestContext = {
            featureName: options.name,
            moduleDir: path.dirname(options.module),
            options,
            host,
        };
        if (options.minimal) {
            options.onlyEmptyRoot = true;
        }
        if (options.skipImport) {
            options.onlyAddFiles = true;
        }
        const fileGeneration = !options.onlyEmptyRoot || (!options.root && options.minimal)
            ? [generateNgrxFilesFromTemplates(options)]
            : [];
        const moduleModification = !options.onlyAddFiles
            ? [
                rules_1.addImportsToModule(requestContext),
                rules_1.addExportsToBarrel(requestContext.options),
            ]
            : [];
        const packageJsonModification = !options.skipPackageJson
            ? [rules_1.addNgRxToPackageJson()]
            : [];
        return schematics_1.chain([
            ...fileGeneration,
            ...moduleModification,
            ...packageJsonModification,
            workspace_1.formatFiles(options),
        ])(host, context);
    };
}
exports.default = generateNgrxCollection;
// ********************************************************
// Internal Function
// ********************************************************
/**
 * Generate 'feature' scaffolding: actions, reducer, effects, interfaces, selectors, facade
 */
function generateNgrxFilesFromTemplates(options) {
    const name = options.name;
    const moduleDir = path.dirname(options.module);
    const excludeFacade = (path) => path.match(/^((?!facade).)*$/);
    const templateSource = schematics_1.apply(schematics_1.url(options.syntax === 'creators' ? './creator-files' : './files'), [
        !options.facade ? schematics_1.filter(excludeFacade) : schematics_1.noop(),
        schematics_1.template(Object.assign(Object.assign(Object.assign({}, options), { tmpl: '' }), devkit_1.names(name))),
        schematics_1.move(moduleDir),
    ]);
    return schematics_1.mergeWith(templateSource);
}
/**
 * Extract the parent 'directory' for the specified
 */
function normalizeOptions(options) {
    return Object.assign(Object.assign({}, options), { directory: devkit_1.names(options.directory).fileName });
}
exports.ngrxGenerator = ngcli_adapter_1.wrapAngularDevkitSchematic('@nrwl/angular', 'ngrx');
