// @nrwl/angular/testing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readFirst = exports.readAll = exports.time = exports.getTestScheduler = exports.hot = exports.cold = void 0;
var jasmine_marbles_1 = require("jasmine-marbles");
Object.defineProperty(exports, "cold", { enumerable: true, get: function () { return jasmine_marbles_1.cold; } });
Object.defineProperty(exports, "hot", { enumerable: true, get: function () { return jasmine_marbles_1.hot; } });
Object.defineProperty(exports, "getTestScheduler", { enumerable: true, get: function () { return jasmine_marbles_1.getTestScheduler; } });
Object.defineProperty(exports, "time", { enumerable: true, get: function () { return jasmine_marbles_1.time; } });
var testing_utils_1 = require("./src/testing-utils");
Object.defineProperty(exports, "readAll", { enumerable: true, get: function () { return testing_utils_1.readAll; } });
Object.defineProperty(exports, "readFirst", { enumerable: true, get: function () { return testing_utils_1.readFirst; } });
