// @nrwl/angular/src/utils/lint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAngularEslintJson = exports.extraEslintDependencies = exports.createAngularProjectESLintLintTarget = void 0;
const workspace_1 = require("@nrwl/workspace");
const versions_1 = require("./versions");
function createAngularProjectESLintLintTarget(projectRoot) {
    return {
        builder: '@nrwl/linter:eslint',
        options: {
            lintFilePatterns: [
                `${projectRoot}/src/**/*.ts`,
                `${projectRoot}/src/**/*.html`,
            ],
        },
    };
}
exports.createAngularProjectESLintLintTarget = createAngularProjectESLintLintTarget;
exports.extraEslintDependencies = {
    dependencies: {},
    devDependencies: {
        '@angular-eslint/eslint-plugin': versions_1.angularEslintVersion,
        '@angular-eslint/eslint-plugin-template': versions_1.angularEslintVersion,
        '@angular-eslint/template-parser': versions_1.angularEslintVersion,
    },
};
const createAngularEslintJson = (projectRoot, prefix) => ({
    overrides: [
        {
            files: ['*.ts'],
            extends: [
                'plugin:@nrwl/nx/angular',
                'plugin:@angular-eslint/template/process-inline-templates',
            ],
            parserOptions: {
                project: [`${projectRoot}/tsconfig.*?.json`],
            },
            rules: {
                '@angular-eslint/directive-selector': [
                    'error',
                    {
                        type: 'attribute',
                        prefix: workspace_1.stringUtils.camelize(prefix),
                        style: 'camelCase',
                    },
                ],
                '@angular-eslint/component-selector': [
                    'error',
                    {
                        type: 'element',
                        prefix: workspace_1.stringUtils.dasherize(prefix),
                        style: 'kebab-case',
                    },
                ],
            },
        },
        {
            files: ['*.html'],
            extends: ['plugin:@nrwl/nx/angular-template'],
            /**
             * Having an empty rules object present makes it more obvious to the user where they would
             * extend things from if they needed to
             */
            rules: {},
        },
    ],
});
exports.createAngularEslintJson = createAngularEslintJson;
