// @nrwl/angular/src/schematics/add-linting/add-linting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLintingGenerator = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
const workspace_1 = require("@nrwl/workspace");
const lint_1 = require("../../utils/lint");
function addLinting(options) {
    return schematics_1.chain([
        workspace_1.addLintFiles(options.projectRoot, "eslint" /* EsLint */, {
            onlyGlobal: false,
            localConfig: lint_1.createAngularEslintJson(options.projectRoot, options.prefix),
            extraPackageDeps: lint_1.extraEslintDependencies,
        }),
        updateProject(options),
    ]);
}
exports.default = addLinting;
function updateProject(options) {
    return (host) => {
        return schematics_1.chain([
            workspace_1.updateJsonInTree(workspace_1.getWorkspacePath(host), (json) => {
                const project = json.projects[options.projectName];
                project.architect.lint = lint_1.createAngularProjectESLintLintTarget(options.projectRoot);
                json.projects[options.projectName] = project;
                return json;
            }),
        ]);
    };
}
exports.addLintingGenerator = ngcli_adapter_1.wrapAngularDevkitSchematic('@nrwl/angular', 'add-linting');
