// @nrwl/angular/src/schematics/library/library.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.libraryGenerator = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const workspace_1 = require("@nrwl/workspace");
const init_1 = require("../init/init");
const add_module_1 = require("./lib/add-module");
const normalize_options_1 = require("./lib/normalize-options");
const update_lib_package_npm_scope_1 = require("./lib/update-lib-package-npm-scope");
const update_project_1 = require("./lib/update-project");
const update_tsconfig_1 = require("./lib/update-tsconfig");
const enable_strict_type_checking_1 = require("./lib/enable-strict-type-checking");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
function default_1(schema) {
    return (host) => {
        const options = normalize_options_1.normalizeOptions(host, schema);
        if (!options.routing && options.lazy) {
            throw new Error(`routing must be set`);
        }
        if (options.enableIvy === true && !options.buildable) {
            throw new Error('enableIvy must only be used with buildable.');
        }
        if (options.publishable === true && !schema.importPath) {
            throw new schematics_1.SchematicsException(`For publishable libs you have to provide a proper "--importPath" which needs to be a valid npm package name (e.g. my-awesome-lib or @myorg/my-lib)`);
        }
        return schematics_1.chain([
            init_1.default(Object.assign(Object.assign({}, options), { skipFormat: true })),
            init_1.addUnitTestRunner(options),
            // TODO: Remove this after Angular 10.1.0
            workspace_1.updateJsonInTree('tsconfig.json', () => ({
                files: [],
                include: [],
                references: [],
            })),
            schematics_1.externalSchematic('@schematics/angular', 'library', {
                name: options.name,
                prefix: options.prefix,
                entryFile: 'index',
                skipPackageJson: !(options.publishable || options.buildable),
                skipTsConfig: true,
            }),
            // TODO: Remove this after Angular 10.1.0
            (host) => {
                host.delete('tsconfig.json');
            },
            schematics_1.move(options.name, options.projectRoot),
            update_project_1.updateProject(options),
            update_tsconfig_1.updateTsConfig(options),
            options.unitTestRunner === 'jest'
                ? schematics_1.externalSchematic('@nrwl/jest', 'jest-project', {
                    project: options.name,
                    setupFile: 'angular',
                    supportTsx: false,
                    skipSerializers: false,
                })
                : schematics_1.noop(),
            options.unitTestRunner === 'karma'
                ? schematics_1.schematic('karma-project', {
                    project: options.name,
                })
                : schematics_1.noop(),
            options.publishable || options.buildable
                ? update_lib_package_npm_scope_1.updateLibPackageNpmScope(options)
                : schematics_1.noop(),
            add_module_1.addModule(options),
            options.strict ? enable_strict_type_checking_1.enableStrictTypeChecking(options) : schematics_1.noop(),
            options.linter !== "none" /* None */ ? addLinting(options) : schematics_1.noop(),
            workspace_1.formatFiles(options),
        ]);
    };
}
exports.default = default_1;
const addLinting = (options) => () => {
    return schematics_1.chain([
        schematics_1.schematic('add-linting', {
            linter: options.linter,
            projectType: 'library',
            projectName: options.name,
            projectRoot: options.projectRoot,
            prefix: options.prefix,
        }),
    ]);
};
exports.libraryGenerator = ngcli_adapter_1.wrapAngularDevkitSchematic('@nrwl/angular', 'library');
