// @nrwl/linter/src/executors/eslint/hasher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fileutils_1 = require("@nrwl/workspace/src/utilities/fileutils");
const app_root_1 = require("@nrwl/workspace/src/utilities/app-root");
const path_1 = require("path");
function run(task, taskGraph, hasher) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (task.overrides['hasTypeAwareRules'] === true) {
            return hasher.hashTaskWithDepsAndContext(task);
        }
        const sources = yield hasher.hashSource(task);
        const deps = allDeps(task.id, taskGraph);
        const nxJson = fileutils_1.readJsonFile(path_1.join(app_root_1.appRootPath, 'nx.json'));
        const tags = deps
            .map((d) => (nxJson.projects[d].tags || []).join('|'))
            .join('|');
        const context = yield hasher.hashContext();
        return {
            value: hasher.hashArray([
                sources,
                tags,
                context.implicitDeps.value,
                context.runtime.value,
            ]),
            details: {
                command: null,
                runtime: context.runtime.runtime,
                implicitDeps: context.implicitDeps.sources,
                sources: { [task.target.project]: sources },
            },
        };
    });
}
exports.default = run;
function allDeps(taskId, taskGraph) {
    return [...taskGraph.dependencies[taskId].map((d) => allDeps(d, taskGraph))];
}
